noir-circuits/blob/src/main.nr

```diff
@@ -186,356 +73,14 @@ fn blob_to_fields__tightly_packed(blob: [F; FIELDS_PER_BLOB]) -> [Field; NOIR_FI
         for j in 1..254 {
             let k = i * 254 + j;
             reconstituted_field *= 2;
-            reconstituted_field += top_bits[k];
-            std::as_witness(reconstituted_field);
+            reconstituted_field += top_bits[k]; // BUG: k is known at compile-time, so this shouldn't be contributing 24k constraints?
+            std::as_witness(reconstituted_field); // BUG: this is costing 4048 gates
```

Comment:
@vezenovm This is falling afoul of the restricted codegen width, right?

Reply:
In the `main` branch, I'm throwing `std::as_witness()` around like throwing darts blindfolded.

I haven't fully grasped when it's actually needed.

I seem to have adopted a rule (without reason) that I should include it for any variable which is declared outside of a `for` loop, and mutated inside the `for` loop. 🤷‍♂️

Reply:
Essentially it should be used when you know an expression can fit inside of the width of the backend you are using and you don't want Noir to create extra intermediate gates. It is a bit hard to determine exactly when `as_witness` is needed and it usually requires a bit of experimentation to determine if Noir is creating an unnecessary intermediate gates. We have made efforts to enable ACIR gen to aware of the backend expression width and the `--bounded-codegen` flag can assist in determining if `as_witness` would be helpful. Although I did run `nargo info --force --bounded-codegen --expression-width=4 ` on this program and I am in fact getting a lower gate count with just `nargo info` 🤔 . I would just not use `as_witness` for now as if it is not used correctly it will add more gates, which looks to be in line with what we are seeing.

Reply:
I'm also seeing an increase of ~2k gates not 4k gates. 30481 (w/o as_witness) -> 32513 (w/ as_witness)

And if you are curious I got 31153 w/ the `--bounded-codegen` flag.